### ISSUE TITLE
Pretraining attribute and object loss don't match the ones in the log file

### ISSUE THREAD

**Issue by kritiagg** [airsplay/lxmert#39]
I tried to load the provided pretrained model with the latest version of the code. But when I tried printing the values for attribute and object, I found a discrepancy between the log file shared for different epochs and the one I got when I printed.
I found the values for the training
obj loss: 2.2379
attr loss : 2.6806

While the one reported in the log file are : Obj: 0.2028 Attr: 0.1267 
I see the same trend in the eval losses as well.

**Comment by airsplay** [comment]
Sorry... I did not fully understand the question. Could you specify the details of "log file"?

**Comment by kritiagg** [comment]
I was comparing the pretrained model you provided with the logfile you provided in another issue: [https://github.com/airsplay/lxmert/issues/22](issue22)
The logfile states that the object loss and attribute loss are 
```
The training loss for Epoch 17 is 2.9553
The losses are Mask_LM: 1.0930 Matched: 0.1697 Obj: 0.1323 Attr: 0.0929 Feat: 0.2087 QA: 1.2587 
Overall Accu 0.3558, gqa Accu 0.4454, visual7w Accu 0.2548, vqa Accu 0.4324, 
The valid loss is 4.5562
The losses are Mask_LM: 1.1740 Matched: 0.2241 Obj: 0.7294 Attr: 0.2575 Feat: 0.2045 QA: 1.9668 
Overall Accu 0.2980, gqa Accu 0.3561, visual7w Accu 0.2016, vqa Accu 0.3650, 
 94%|██████████████████████████████████████████████████████████▏   | 43899/46742 [18:11:36<1:11:55,  1.52s/it
```
But if I load the pretrained model provided by you and try to print the same losses, I get very high values for the object and attribute losses. I get obj loss: 2.2379
attr loss : 2.6806 for the training losses.

**Comment by airsplay** [comment]
Sorry. I am not sure about this since it is not observed by my experiments and is not reported elsewhere.

I am currently not super clear about how you "load the pretrained model provided by you and try to print the same losses". If you are willing to provide the detailed instructions that you get these results, I could test it locally. Hope it helps!

**Comment by marsuplink** [comment]
I'm also seeing something similar, the model predicted categories for objects and attributes were overwhelmingly monotonous, i.e. the predicted classes fall under just one or two classes, causing the object and attribute losses to be high. Have you ever encountered this issue of the same classes being predicted over and over?

Thanks!

**Comment by kritiagg** [comment]
Hi @airsplay 

Here is the steps I followed: 
- Downloaded the code and data for pretraining and finetuning as mentioned in your repo.
- Downloaded the [pretrained model](http://nlp.cs.unc.edu/data/model_LXRT.pth) and put it in the snap/pretrain/pretrained/
- Changed the learning rate in the run/lxmert_pretrain.sh script to 0.0. 
`
name=lxmert


output=snap/pretrain/$name
mkdir -p $output/src
cp -r src/* $output/src/
cp $0 $output/run.bash


CUDA_VISIBLE_DEVICES=$1 PYTHONPATH=$PYTHONPATH:./src \
    python src/pretrain/lxmert_pretrain.py \
    --taskMaskLM --taskObjPredict --taskMatched --taskQA \
    --visualLosses obj,attr,feat \
    --wordMaskRate 0.15 --objMaskRate 0.15 \
    --train mscoco_train,mscoco_nominival,vgnococo --valid mscoco_minival \
    --llayers 9 --xlayers 5 --rlayers 5 \
    --fromScratch \
    --batchSize 64 --optim bert --lr 0.0 --epochs 20 \
    --tqdm --output $output ${@:2}
`
- Added a set trace on line 294 in src/pretrain/lxmert_pretrain.py  <to see the visual losses: obj, attr,  feat> 
- Ran the command: 
`bash run/lxmert_pretrain.bash 0,1,2,3 --multiGPU --tiny --loadLXMERT snap/pretrained/model`
- And I get the result: 
tensor([1.3126, 0.5437, 4.8402, 2.7473, 0.2967, 7.2280])
- I was comparing the these losses with the ones you provided in another issue: https://github.com/airsplay/lxmert/issues/22
The logfile states that the object loss and attribute loss are 
Mask_LM: 1.0930 Matched: 0.1697 Obj: 0.1323 Attr: 0.0929 Feat: 0.2087 QA: 1.2587 

`The training loss for Epoch 17 is 2.9553
The losses are Mask_LM: 1.0930 Matched: 0.1697 Obj: 0.1323 Attr: 0.0929 Feat: 0.2087 QA: 1.2587 
Overall Accu 0.3558, gqa Accu 0.4454, visual7w Accu 0.2548, vqa Accu 0.4324, 
The valid loss is 4.5562
The losses are Mask_LM: 1.1740 Matched: 0.2241 Obj: 0.7294 Attr: 0.2575 Feat: 0.2045 QA: 1.9668 
Overall Accu 0.2980, gqa Accu 0.3561, visual7w Accu 0.2016, vqa Accu 0.3650, 
 94%|██████████████████████████████████████████████████████████▏   | 43899/46742 [18:11:36<1:11:55,  1.52s/it`
- As you can see the Matched loss, obj loss and attr losses are very high for the loaded model. Although the the others are still within the range. (QA loss is also high but that is because I am not loading the QA head) And I understand the losses have some noise since they were calculated only for a batch but they should still be in the range of the actual losses. 

I hope this clarifies the question I was asking?

**Comment by airsplay** [comment]
Thanks for these clarifications! 

I got the reason now. In short, the number that you printed is the **training loss** but I showed the **evaluation loss** instead. 

The **evaluation loss** is tested under the  condition `model.eval` here:
https://github.com/airsplay/lxmert/blob/827087bbf7b8eb4cb2b648aeb1875ded79bf4ff0/src/pretrain/lxmert_pretrain.py#L361
It thus gets rid of the randomness provided by the dropout layers. 

Another thing to be noticed is that your provided loss is only the loss of a batch. A more accurate evaluation loss should be averaged over the corpus. It is solved by the function `evaluate_epoch` here: https://github.com/airsplay/lxmert/blob/827087bbf7b8eb4cb2b648aeb1875ded79bf4ff0/src/pretrain/lxmert_pretrain.py#L360

**Comment by kritiagg** [comment]
Hi I was only comparing the train losses between the one from the logfile and the one I printed. 

But to further bring the point home, I have also printed the logfile with settings as mentioned in my earlier comment. Here is what I get: < I print both the eval loss and the train loss and for both obj and attr losses are higher>
```
The losses are Mask_LM: 1.2365 Matched: 0.3485 Obj: 5.3129 Attr: 3.0935 Feat: 0.2901 QA: 8.9026
Overall Accu 0.0000, gqa Accu 0.0001, visual7w Accu 0.0000, vqa Accu 0.0000,
The valid loss is 18.8215
The losses are Mask_LM: 1.2626 Matched: 0.5132 Obj: 4.2496 Attr: 3.0363 Feat: 0.2626 QA: 9.4971
Overall Accu 0.0000, gqa Accu 0.0000, visual7w Accu 0.0000, vqa Accu 0.0000,
                                                         | 1759/5246 [14:40<27:56,  2.08it/s]
```

**Comment by airsplay** [comment]
Thanks for the clarification. 

I think that it might because the command (as you provided above) does not load the pre-trained mode files. 
```
name=lxmert
output=snap/pretrain/$name
mkdir -p $output/src
cp -r src/* $output/src/
cp $0 $output/run.bash

CUDA_VISIBLE_DEVICES=$1 PYTHONPATH=$PYTHONPATH:./src
python src/pretrain/lxmert_pretrain.py
--taskMaskLM --taskObjPredict --taskMatched --taskQA
--visualLosses obj,attr,feat
--wordMaskRate 0.15 --objMaskRate 0.15
--train mscoco_train,mscoco_nominival,vgnococo --valid mscoco_minival
--llayers 9 --xlayers 5 --rlayers 5
*****--fromScratch******
--batchSize 64 --optim bert --lr 0.0 --epochs 20
--tqdm --output $output ${@:2}
```

The command uses option `--fromScratch`, which means that the model will randomly intialize the weight. In order to load the weight, please use `--load /path/to/pre-trained/model`.

**Comment by airsplay** [comment]
> I'm also seeing something similar, the model predicted categories for objects and attributes were overwhelmingly monotonous, i.e. the predicted classes fall under just one or two classes, causing the object and attribute losses to be high. Have you ever encountered this issue of the same classes being predicted over and over?
> 
> Thanks!

Thanks. I have not checked the predicted results.  It might be related to a bug found here: https://github.com/airsplay/lxmert/issues/41. I will test it when I got enough computational resources in my hand.

**Comment by kritiagg** [comment]
Hi @airsplay 
I used the command : <as mentioned in my  [comment](https://github.com/airsplay/lxmert/issues/39#issuecomment-589471987) >
`bash run/lxmert_pretrain.bash 0,1,2,3 --multiGPU --tiny --loadLXMERT snap/pretrained/model`

And That loads the model properly, since Masked LM loss and feature regression are similar to the one you provided in the log file.

**Comment by airsplay** [comment]
Oh, sorry about this misunderstanding. I will check it when I got some time and GPUs.

**Comment by airsplay** [comment]
Thanks a lot to point it out. I got the reason. The released pre-training is trained with a different protocol thus need a few-step adaption to this code.  

To get the same results as the log and for directly study of the pre-training, please use the new snapshots `https://nlp.cs.unc.edu/data/github_pretrain/lxmert20/Epoch20_LXRT.pth` provided here: https://github.com/airsplay/lxmert/issues/22#issuecomment-543364509.

Sampled loss:
`tesor([[0.9515, 0.2117, 0.0970, 0.0616, 0.1982, 9.2852]])`

Note: the last loss is still high because the option `--loadLXMERT` instead of `--load` is used, which re-initialize the answer head.
